optuna/storages/_journal/file.py

```diff
@@ -162,7 +162,7 @@ def __init__(self, file_path: str, lock_obj: Optional[JournalFileBaseLock] = Non
 
     def read_logs(self, log_number_from: int) -> List[Dict[str, Any]]:
         logs = []
-        with open(self._file_path, "r") as f:
+        with open(self._file_path, "rb") as f:
```

Comment:
According to the reference of [`io.IOBase.readline`](https://docs.python.org/3/library/io.html#io.IOBase.readline), `f.readline()` split the lines by `\n`.

I'm a bit concerned the implementation relying on this specification because 
1) Some platforms (e.g., old macos) use `\r` as line delimiter.
2) I'm not sure if this behavior is consistent with different Python interpreter.

> The line terminator is always b'\n' for binary files; for text files, the newline argument to [open()](https://docs.python.org/3/library/functions.html#open) can be used to select the line terminator(s) recognized.

How about setting `newline="\n"` to `open` to read and write log files?
```suggestion
        with open(self._file_path, "r", newline="\n") as f:
```

And also, L193 will be as follows:
```diff
- with open(self._file_path, "a", encoding="utf-8") as f:
+ with open(self._file_path, "a", encoding="utf-8", newline="\n") as f:
```

As a positive side effect, the log file can be Windows and Linux compatible.

Reply:
Thank you for the suggestion.

In my recognition, the log file is fully compatible between Windows and Linux in the current implementation since we use binary mode and explicitly specify "\n" as a newline character. So it seems to work even if `f.readline()` splits lines with `\n` in binary files. I didn't test on Windows though. 

https://github.com/optuna/optuna/blob/997bd041e239486febb7479db359a1919ca43fe5/optuna/storages/_journal/file.py#L192

Also as a positive side effect on binary-mode, we can get rid of `.encode("utf-8")` as @contramundum53 suggested.

What do you think?

Reply:
But on second thought, given that my logic, we should also use binary-mode in `append_logs`.

Reply:
> But on second thought, given that my logic, we should also use binary-mode in `append_logs`.

Agree. I am also concerned that the current PR uses both text and binary modes. Unexpected side effects may occur, and it may be safe to choose either one, and control the line delimiter.

Reply:
> I would say it's safer to use a binary mode in `read_logs` since we need to calculate the byte offset to seek the file position.

That makes sense. Your first suggestion looks solid to me.
Could you update the PR, please?

Nit-picky, but could you update the following flag of `open` as well?
I don't think it changes the behavior, but just for consistency.

```diff
--- a/optuna/storages/_journal/file.py
+++ b/optuna/storages/_journal/file.py
@@ -157,7 +157,7 @@ class JournalFileStorage(BaseJournalLogStorage):
     def __init__(self, file_path: str, lock_obj: Optional[JournalFileBaseLock] = None) -> None:
         self._file_path: str = file_path
         self._lock = lock_obj or JournalFileSymlinkLock(self._file_path)
-        open(self._file_path, "a").close()  # Create a file if it does not exist
+        open(self._file_path, "ab").close()  # Create a file if it does not exist
```

Reply:
I confirmed that the following change worked on my Windows PC.

```diff
diff --git a/optuna/storages/_journal/file.py b/optuna/storages/_journal/file.py
index b3c98a8ea..8e26bd43b 100644
--- a/optuna/storages/_journal/file.py
+++ b/optuna/storages/_journal/file.py
@@ -157,7 +157,7 @@ class JournalFileStorage(BaseJournalLogStorage):
     def __init__(self, file_path: str, lock_obj: Optional[JournalFileBaseLock] = None) -> None:
         self._file_path: str = file_path
         self._lock = lock_obj or JournalFileSymlinkLock(self._file_path)
-        open(self._file_path, "a").close()  # Create a file if it does not exist
+        open(self._file_path, "ab").close()  # Create a file if it does not exist
         self._log_number_offset: Dict[int, int] = {0: 0}

     def read_logs(self, log_number_from: int) -> List[Dict[str, Any]]:
@@ -190,7 +190,7 @@ class JournalFileStorage(BaseJournalLogStorage):
     def append_logs(self, logs: List[Dict[str, Any]]) -> None:
         with get_lock_file(self._lock):
             what_to_write = "\n".join([json.dumps(log) for log in logs]) + "\n"
-            with open(self._file_path, "a", encoding="utf-8") as f:
-                f.write(what_to_write)
+            with open(self._file_path, "ab") as f:
+                f.write(what_to_write.encode("utf-8"))
                 f.flush()
                 os.fsync(f.fileno())
```

Reply:
Thank you for your review! I applied the change and pushed it at 8f1117947.

Reply:
Thanks!